### PR TITLE
Add the capability to specify JIT options on the SPMI command line.

### PR DESCRIPTION
**Pull request by pgavlin**
This change adds two new flags, `-jitoption` and `-jit2option`, to the
SuperPMI command line. These flags are cumulative, and are used to
specify JIT config options (e.g. `COMPlus_JitDisasm`, albeit without the
`COMPlus_` prefix) to the base and diff JITs, respectively.

This change also removes Smarty-specific code and support for capturing
and replaying environment variables. The former is no longer terribly
useful and the latter has been obviated by the JIT host interface.

### CODE REVIEW

**Comment by pgavlin** [comment]
@BruceForstall @sandreenko PTAL

File: src/ToolBox/superpmi/superpmi-shared/methodcontext.cpp [hunk 1]
```diff
@@ -498,128 +498,8 @@ void MethodContext::recGlobalContext(const MethodContext& other)
...
-        while (*l_EnvStr != '\0')
-            l_EnvStr++;
-        l_EnvStr++;
-    }
-    FreeEnvironmentStringsA(l_val);
-}
 void MethodContext::dmpEnvironment(DWORD key, const Agnostic_Environment& value)
 {
-    printf("Environment key %u, value '%s' '%s'", key, (LPCSTR)Environment->GetBuffer(value.name_index),
-           (LPCSTR)Environment->GetBuffer(value.val_index));
```
Comment (BruceForstall) [inline src/ToolBox/superpmi/superpmi-shared/methodcontext.cpp]: Why did you leave an empty MethodContext::dmpEnvironment()?

File: src/ToolBox/superpmi/superpmi/commandline.cpp [hunk 2]
```diff
@@ -113,6 +114,12 @@ void CommandLine::DumpHelp(const char* program)
     printf("     Use disassembly tools from the CoreDisTools library\n");
     printf("\n");
 #endif // USE_COREDISTOOLS
+    printf(" -jitoption key=value\n");
+    printf("     Set the JIT option named \"key\" to \"value\" for JIT 1. NOTE: do not use a \"COMPlus_\" prefix!\n");
+    printf("\n");
+    printf(" -jit2option key=value\n");
+    printf("     Set the JIT option named \"key\" to \"value\" for JIT 2. NOTE: do not use a \"COMPlus_\" prefix!\n");
+    printf("\n");
```
Comment (BruceForstall) [inline src/ToolBox/superpmi/superpmi/commandline.cpp]: Do we really want the user model to allow specifying different flags to each JIT? Would you really want to pass something different to baseline and diff? Wouldn't it be inconvenient to have to duplicate what you are specifying twice?

**Comment by BruceForstall** [comment]
I assume there are no asm diffs? (To prove the environment was not used)

**Comment by BruceForstall** [comment]
It seems a shame to lose the mapping from MethodContext to `BVT_TEST_ID`. I can't say I'd used that more than once or twice, but it's the only real way to map back from a failure in replay to an actual test case.

**Comment by sandreenko** [comment]
Is it safe to change all JitFlags between a collection and a replay? I think it can lead to the wrong results. 

Maybe we should enable possibility to set only  dump flags?

File: src/ToolBox/superpmi/superpmi-shared/methodcontext.cpp (hunk 1 above)
Comment (pgavlin) [inline src/ToolBox/superpmi/superpmi-shared/methodcontext.cpp]: I didn't want to invalidate existing MCH files, so I left the LWM definition for `Agnostic_Environment`. The macro that is used to define LWM definitions requires that a `dup` method exists.

I will add a comment to this effect.

File: src/ToolBox/superpmi/superpmi/commandline.cpp (hunk 2 above)
Comment (pgavlin) [inline src/ToolBox/superpmi/superpmi/commandline.cpp]: The ability to specify different options to each JIT is the crux of this change: it allows a user to use the same JIT with different options for the base and the diff. This allows us to more easily experiment with JIT behavior that is governable by switches.

**Comment by pgavlin** [comment]
> I assume there are no asm diffs? (To prove the environment was not used)

I haven't checked; that's a good suggestion.

> It seems a shame to lose the mapping from MethodContext to BVT_TEST_ID. I can't say I'd used that more than once or twice, but it's the only real way to map back from a failure in replay to an actual test case.

Noted. It shouldn't be too hard to restore if we find it useful in the future, but I don't think it's carrying its own weight at the moment (and I don't think it's worth keeping the environment bits just to serve this purpose).

> Is it safe to change all JitFlags between a collection and a replay? I think it can lead to the wrong results.

Why do you think that it can lead to wrong results? It could certainly lead to failures during replay if a flag causes the JIT to ask for data that we did not collect, but that's okay. This feature is primarily intended to enable experiments in which the same JIT is used as both the base and the diff JIT but with different config flags.

**Comment by sandreenko** [comment]
> Why do you think that it can lead to wrong results? It could certainly lead to failures during replay if a flag causes the JIT to ask for data that we did not collect, but that's okay. This feature is primarily intended to enable experiments in which the same JIT is used as both the base and the diff JIT but with different config flags.

I think that it is possible for jit asks the same questions, but expect different answers depending on JitFlags. It means we will have successful replay, but it will not show the real picture, However, it is already can happen with collisions in spmi maps, so I think we are fine.

File: src/ToolBox/superpmi/superpmi-shared/methodcontext.cpp (hunk 1 above)
Comment (sandreenko) [inline src/ToolBox/superpmi/superpmi-shared/methodcontext.cpp]: Does it mean that we can safely delete it with a next breaking change?

**Comment by BruceForstall** [comment]
> I think that it is possible for jit asks the same questions, but expect different answers depending on JitFlags. 

That would mean that the VM adjusts the answers based on the JitFlags? I don't think that happens.

**Comment by karelz** [comment]
What's status of this PR? No update for last 4 weeks.

**Comment by pgavlin** [comment]
@BruceForstall @sandreenko PTAL

**Comment by BruceForstall** [comment]
@pgavlin Are you going to fix the conflict (and retrigger CI runs)?

**Comment by pgavlin** [comment]
Definitely--just wanted to collect any additional feedback that I could fix while doing so.

File: src/ToolBox/superpmi/superpmi-shim-collector/icorjitcompiler.cpp [hunk 3]
```diff
@@ -63,9 +63,6 @@ CorJitResult __stdcall interceptor_ICJC::compileMethod(ICorJitInfo*
         our_ICorJitInfo.mc->recGlobalContext(*g_globalContext);
     }
 
-    // Record a simple view of the environment
-    our_ICorJitInfo.mc->recEnvironment();
```
Comment (sandreenko) [inline src/ToolBox/superpmi/superpmi-shim-collector/icorjitcompiler.cpp]: This call dumped environment variables during the collection, how was it replaced?

Comment (pgavlin) [reply src/ToolBox/superpmi/superpmi-shim-collector/icorjitcompiler.cpp]: It wasn't. We could certainly add the capability to dump the JIT config settings during/after a collection, though.

Comment (sandreenko) [reply src/ToolBox/superpmi/superpmi-shim-collector/icorjitcompiler.cpp]: During SuperPMI collection some tests set flags as complus_jitstress, complus_enableavx and etc. SpmiShim record them and the spmi read them from mc file and apply for the correct replay. It is critical to keep this mechanism working.

Comment (pgavlin) [reply src/ToolBox/superpmi/superpmi-shim-collector/icorjitcompiler.cpp]: Ah, I misunderstood your question. That mechanism will continue to work: those settings will be captured by the JitHost.

Comment (sandreenko) [reply src/ToolBox/superpmi/superpmi-shim-collector/icorjitcompiler.cpp]: Cool, now I see it.

**Comment by BruceForstall** [comment]
Presumably we could still save the `BVT_TEST_ID`, but just in the JitHost data instead of the Environment map? And thus still maintain that functionality? I guess I'm not wedded to it, though.

**Comment by karelz** [comment]
@pgavlin looks like the PR needs merge & CI leg failures attention ...

**Comment by pgavlin** [comment]
@dotnet-bit test this please

**Comment by pgavlin** [comment]
Looks like there's some sort of massive CI backup. @mmitche any ideas?

**Comment by mmitche** [comment]
@pgavlin Node stopped connecting to Jenkins last night (reason unknown at moment). I think I may have resolved the immediate issue, but there are still odd exceptions coming from Azure.  Please stand by

**Comment by pgavlin** [comment]
@dotnet-bot 
test OSX10.12 x64 Checked Build and Test
test Ubuntu x64 Checked Build and Test

**Pull request merged**